File: Easy/IntersectionOfLinkedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type headA, headB: ListNode
        :rtype: ListNode
        """
        if not headA or not headB:
            return None
        a_pointer, b_pointer = headA, headB
        a_list_len, b_list_len = 0, 0
        while a_pointer.next is not None:
            a_pointer = a_pointer.next
            a_list_len += 1
        while b_pointer.next is not None:
            b_pointer = b_pointer.next
            b_list_len += 1
        if a_pointer is b_pointer:  # we have a merge
            diff = abs(a_list_len - b_list_len)
            counter = 0
            a_pointer, b_pointer = headA, headB
            while counter < diff:
                if a_list_len > b_list_len:
                    a_pointer = a_pointer.next
                    counter += 1
                else:
                    b_pointer = b_pointer.next
                    counter += 1
            while a_pointer is not b_pointer:
                a_pointer, b_pointer = a_pointer.next, b_pointer.next
            return a_pointer
        return None

File: Medium/intersectionOfLinkedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
  if (!headA || !headB) {
    return null
  }
  let aNode = headA
  let bNode = headB
  let leaderNode
  let followerNode

  while (aNode || bNode) {
    if (aNode === bNode) {
      return aNode
    }
    if (!aNode || !bNode) {
      if (!leaderNode) {
        leaderNode = aNode ? headA : headB
        followerNode = aNode ? headB : headA
      }
      leaderNode = leaderNode.next
    }
    aNode = aNode ? aNode.next : null
    bNode = bNode ? bNode.next : null
  }

  while (leaderNode !== followerNode && leaderNode) {
    leaderNode = leaderNode.next
    followerNode = followerNode.next
  }

  return leaderNode
}

// Another way to do this.

var getIntersectionNode = function(headA, headB) {
  var pointerA = headA
  var pointerB = headB

  while (pointerA !== pointerB) {
    pointerA = pointerA === null ? headB : pointerA.next
    pointerB = pointerB === null ? headA : pointerB.next
  }

  return pointerA
}
